Adding SSO CILogon

lagoan opened the pull request:
This PR includes the combined efforts to add SSO to DMP Assistant.

Will be cleaning the PR and submit at a later time

Commit:
Adding a openid connect gem

Commit:
Omniauth Callback changes, this is not finalized will be changing

200455939-yashu committed:
adding cilogon to the devise settings and user changes, this is not finalized

200455939-yashu committed:
adding sign-in with the cilogon option

200455939-yashu committed:
adding the openid devise settings

200455939-yashu committed:
adding cilogon callback to the route

200455939-yashu committed:
adding the session changes to the cilogon, not finalized

200455939-yashu committed:
session changes to support cilogon

200455939-yashu committed:
changes not required will be cleaned

200455939-yashu committed:
Merge remote-tracking branch 'origin/integration' into yashu-sso-cilogon

200455939-yashu committed:
Temporary changes to test the credentials

200455939-yashu committed:
This file contains multiple trial and error methods on callback needs more modifications

200455939-yashu committed:
since we will be having the eppn value with cilogon instead of the uid

200455939-yashu committed:
Devise openid_connect trial and errors

200455939-yashu committed:
merge after resolving gem lock file conflicts by accepting the incoming changes

200455939-yashu committed:
This is working code so far! still need lot of clean-up on this

200455939-yashu committed:
Merge remote-tracking branch 'origin/integration' into yashu-sso-cilogon

200455939-yashu committed:
The debugging of the sessions. This major Needs cleanup

200455939-yashu committed:
Altering the MAX cookies size for the browser

200455939-yashu committed:
To enable the recaptcha in localhost we use 127 instead of localhost

200455939-yashu committed:
This the working code of the SSO sing-in testing connect to ORCID

200455939-yashu committed:
Merge remote-tracking branch 'origin/integration' into yashu-sso-cilogon

lagoan committed:
WIP Updates to sso

Adding some missing methods and start cleanup.

We still need to clean up and fix the `openid_connect` method on
`omniauth_callbacks_controller.rb`.

200455939-yashu committed:
code clean up and added test cases

lagoan committed:
Add omniauth tests and fix identifier_scheme bug

Here we have a first pass of working integration tests for omniauth SSO.

The tests still need to check for the messages we are setting as
notifications.

We can think about splitting the tests into integration and controller
tests.

lagoan committed:
Cleanup of integration tests

200455939-yashu committed:
Testcases in progress changes

200455939-yashu committed:
Merging the branches

200455939-yashu committed:
Merging the branches

200455939-yashu committed:
WIP of testcases

200455939-yashu committed:
adding the keys secrets file

200455939-yashu committed:
reupdating the return on empty email

200455939-yashu committed:
Removing commented byebugs

200455939-yashu committed:
Test cases with flash one is clearing checking the other one

200455939-yashu committed:
fixes with button name and uri updates

200455939-yashu committed:
syntax correction

200455939-yashu committed:
updates with URI url and testcase link with

lagoan committed:
Add link account with CILogon

This change adds the behaviour to link authenticating accounts from
CILogon.

The test is missing from the integration side.

lagoan committed:
Add missing changes

lagoan committed:
Merge branch 'yashu-sso-cilogon' into yashu-sso-link-accounts

lagoan committed:
Fix failing tests due to incorrect mocking

lagoan committed:
Add extended scope for SSO

lagoan committed:
Update devise initializer

lagoan committed:
Add another devise fix

lagoan committed:
Pass tests for openid_connect_sso_test

This commit helps the tests pass but we are not running them in the
browser and we are not testing for flash messages.

We are extrapolating from the accounts being created.

lagoan committed:
Return maximum name for identifier scheme

This is a value that was set by upstream.

200455939-yashu committed:
Testcases for the Ominiauth controller openid connect

200455939-yashu committed:
Code cleanup - specs controllers omniauth

lagoan committed:
Robust name handling when creating user & CC

This commit handles the possible lack of name information from the IDP.

Also clean up a bit of code

lagoan commented on app/controllers/application_controller.rb:
```diff
@@ -161,6 +161,17 @@ def after_sign_out_path_for(resource_or_scope)
   end
   # -------------------------------------------------------------
 
+  ##
+  # Sign out of cilogon  SSO local session too.
+  # # -------------------------------------------------------------
+  # def after_sign_out_path_for(resource_or_scope)
+  #   url = "#{Rails.configuration.x.openid_connect&.logout_url}#{root_url}"
+  #   return url if Rails.configuration.x.openid_connect&.enabled
+
+  #   super
+  # end
+  # # -------------------------------------------------------------
+
```
Will be removing this

lagoan commented on app/controllers/orgs_controller.rb:
```diff
@@ -5,7 +5,7 @@ class OrgsController < ApplicationController
 
   after_action :verify_authorized, except: %w[
-    shibboleth_ds shibboleth_ds_passthru search
+    shibboleth_ds shibboleth_ds_passthru search #cilogon_ds cilogon_ds_passthru
```
Removing this

lagoan committed:
More code cleanup

lagoan committed:
Code cleanup

lagoan commented on app/views/plans/new.html.erb:
```diff
@@ -117,7 +117,6 @@
           </div>
           <div class="col-xs-6">
             <span id="multiple-templates">
-              <%= _('We found multiple DMP templates corresponding to your primary research organisation') %>
```
@200455939-yashu Why did we decide to remove this line ?

lagoan committed:
Code cleanup

lagoan commented on spec/controllers/omniauth_callbacks_controller_spec.rb:
```diff
@@ -0,0 +1,60 @@
+require 'rails_helper'
+require 'byebug'
+
+# RSpec.describe Users::OmniauthCallbacksController, type: :controller do
```
Commented out these test as we are merging this work from another branch. @200455939-yashu

200455939-yashu replied:
this is resolved in the PR #869

lagoan commented on spec/integration/openid_connect_sso_test.rb:
```diff
@@ -0,0 +1,56 @@
+      expect(identifiable.surname).to_not eql('Doe')
+
+      # XXX Check for flash notice message linked successfully
+      # expect(page).to have_content('Linked succesfully')
```
We will need to test for this flash notice at another time

aaronskiba replied:
For it to run during workflows, I think `spec/integration/openid_connect_sso_test.rb` needs to be renamed to `spec/integration/openid_connect_sso_spec.rb`. https://github.com/rspec/rspec-rails?tab=readme-ov-file#running-specs

lagoan committed:
Code cleanup

aaronskiba commented on config/secrets.yml:
```diff
@@ -14,6 +14,7 @@ test:
   database_test_url: <%= ENV['DATABASE_TEST_URL'] %>
   devise_pepper: <%= ENV['DEVISE_PEPPER'] %>
   devise_secret_key: <%= ENV['DEVISE_SECRET_KEY'] %>
+  dmproadmap_host: <%= ENV['DMPROADMAP_HOST'] %>
```
`dmproadmap_host: <%= ENV['DMPROADMAP_HOST'] %>` can be removed from test and development. We won't be using it for devise config's redirect_uri.

lagoan replied:
You are right, making changes

aaronskiba commented on public/tinymce/skins/oxide/content.css:
```diff
@@ -243,6 +243,16 @@ div.mce-footnotes li > a.mce-footnotes-backlink {
     display: none;
```
`public/tinymce/skins/` is auto-generated whenever the `rails assets:precompile` is executed. It should probably be added to `.gitignore` (same should be done upstream). It would be okay to leave these here for now, but I'll make a ticket.

lagoan committed:
Add missing line in new plans view

200455939-yashu committed:
Issue fix for the multiple accounts

aaronskiba commented on config/database.yml:
```diff
@@ -15,6 +15,9 @@ development:
   <<: *defaults
+  username: <%= ENV['DATABASE_USER'] %>
+  password: <%= ENV['DATABASE_PASSWORD'] %>
+  host: <%= ENV['DATABASE_URL'] || '127.0.0.1' %>
```
I don't think code changes to `config/database.yml` are needed. `<<: *defaults` should already be doing what is added here.

lagoan replied:
👍🏽

aaronskiba reviewed (changes requested):
In terms of reviewing, I've only done a quick scan so far.
In addition to these comments, I can see that the PostgreSQL and the Rubocop GitHub actions are failing.
 - The PostgreSQL failures related to `./spec/features/plans/exports_spec.rb` can be ignored, https://github.com/portagenetwork/roadmap/issues/870 addresses them.
 - The RuboCop offences should be addressed.

lagoan committed:
Clean up configuration

200455939-yashu committed:
Resolving conflicts after the yashu-sso-link-accounts

200455939-yashu committed:
Merge remote-tracking branch 'origin/yashu-sso-2user-accounts-issue-fix' into yashu-controller-spec

aaronskiba commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -12,6 +12,55 @@ class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+                                                      attrs: auth,
+                                                      identifiable: current_user)
+
+        flash[:notice] = 'Linked succesfully'
```
Add `_()` to `'Linked succesfully'` for translation.io

aaronskiba commented on app/views/devise/registrations/_external_openid_connect.html.erb:
```diff
@@ -0,0 +1,21 @@
+                    'data-toggle': "tooltip", method: :post %>
+<% else %>
+<%# If We do have and id we need to present the option to unlink %>
+<% unlinktext = _("Unlink your account from #{scheme.description}. You can link again at any time.") %>
```
I've had trouble getting translatable strings to work in the past.
`_("Unlink your account from #{scheme.description}. You can link again at any time.")` may have to be changed to the following:
`_("Unlink your account from %{scheme_description}. You can link again at any time.") % { scheme_description: scheme.description}`

The same issue exists on line 12.

aaronskiba replied:
There are several strings needing this update (including https://github.com/portagenetwork/roadmap/pull/875/files)

200455939-yashu committed:
Resolving the conflicts after merge

lagoan committed:
Code cleanup

aaronskiba commented on app/views/shared/_sign_in_form.html.erb:
```diff
@@ -34,4 +34,17 @@
+      <p class="text-center fontsize-h4">- <%= _('or') %> -</p>
+      <div class="form-group">
+        <span class="center-block btn-group-justified">
+          <%= link_to "Sign in with CILogon", user_openid_connect_omniauth_authorize_path, method: :post, data: { turbo: false }, class: 'btn btn-default' %>
```
Change to _('Sign in with CILogon')

lagoan committed:
Add mising translation piece

lagoan committed:
Code cleanup

200455939-yashu committed:
adding test cases for the linked successfylly and 2 users condition

200455939-yashu committed:
Translation related changes

200455939-yashu committed:
Merge remote-tracking branch 'origin/yashu-sso-link-accounts' into yashu-controller-spec

200455939-yashu committed:
Spelling correction

200455939-yashu committed:
Removing the byebug and the updates related to translations.

lagoan committed:
Add CHANGELOG entry

200455939-yashu committed:
Adding the changelog

200455939-yashu committed:
commit after Resolving the conflicts in Changelog

aaronskiba commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -12,6 +12,57 @@ class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+
+      if auth.info.email.nil? && user.nil?
+        # If email is missing we need to request the user to register with DMP.
+        # User email can be missing if the usFFvate or trusted clients only we won't get the value.
```
I don't quite understand this sentence (line 25).

lagoan replied:
@200455939-yashu I think you addef this comment. Could you elaborate ?

aaronskiba replied:
Will be unable to sort this out in time. Marking it as resolved so that the PR can be merged into integration.

aaronskiba commented on app/models/user.rb:
```diff
@@ -182,6 +182,26 @@ def self.from_omniauth(auth)
+
+    return user if user
+
+    User.create!(
+      firstname: provider_data.info&.first_name.present? ? provider_data.info.first_name : _('First name'),
+      surname: provider_data.info&.last_name.present? ? provider_data.info.last_name : _('Last name'),
+      email: provider_data.info.email,
+      # We don't know which organization to setup so we will use other
+      org: Org.find_by(is_other: true),
+      accept_terms: true,
+      password: Devise.friendly_token[0, 20]
+    )
+  end
```
At a later date, it might be best to have the user fill this out before, rather than after creating an account via SSO (i.e. choose an `Org`, manually accept the terms, etc.). Or maybe we could have another flash message informing them that they should update these default values?

lagoan replied:
You are right, a flash notice will be good to let them know if they need to update their name.

aaronskiba replied:
I created the following issue: https://github.com/portagenetwork/roadmap/issues/889 Marking this as resolved.

aaronskiba commented on config/initializers/devise.rb:
```diff
@@ -260,6 +261,7 @@
   # Any entries here MUST match a corresponding entry in the identifier_schemes table as
   # well as an identifier_schemes.schemes section in each locale file!
   OmniAuth.config.full_host = Rails.application.secrets.omniauth_full_host
+  OmniAuth.config.silence_get_warning = true
```
We might not need `OmniAuth.config.silence_get_warning = true` (because we don't have `OmniAuth.config.allowed_request_methods = %i[get post]`).

According to https://github.com/omniauth/omniauth/releases/tag/v2.0.0:

```
GET
Because using GET for login poses concerns (see [OWASP Cheatsheet](https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#cross-site-request-forgery-prevention-cheat-sheet)), after upgrading OmniAuth will log a :warn level log with every GET request to a login path when your OmniAuth.config.allowed_request_methods include :get.

If you have a workflow that absolutely requires you to use GET for the request_phase, you can disable this warning using

OmniAuth.config.silence_get_warning = true
It is very important that you do not do this just to circumvent having to change your inputs or login flow, as using GET for most auth workflows is not what you want.
```

200455939-yashu committed:
Review Changes

lagoan committed:
Fix tests

200455939-yashu committed:
Merge branch 'yashu-sso-link-accounts' into yashu-controller-spec

lagoan committed:
Merge branch 'integration' into yashu-sso-link-accounts

200455939-yashu committed:
About and help page updates of SSO

200455939-yashu committed:
Merge branch 'yashu-sso-link-accounts' into yashu-controller-spec

200455939-yashu committed:
PR update on changelog

lagoan committed:
Update CHANGELOG entry

lagoan committed:
Merge branch 'yashu-sso-link-accounts' from repo

aaronskiba commented on app/views/devise/registrations/_personal_details.html.erb:
```diff
@@ -66,30 +66,22 @@
-      <% if scheme.name.downcase == 'shibboleth' %>
+  <div class="form-group col-xs-8">
         <label class='control-label'>
           <span class="aria-only" aria-hidden="false"><%= _('Institutional credentials') %></span>
```
We talked about how the linked account can be non-institutional. Do we want to update `_('Institutional credentials')`?

lagoan replied:
We will keep this as is for the time being.

200455939-yashu committed:
Merge branch 'yashu-sso-link-accounts' into yashu-sso-about-help-updates

aaronskiba commented on app/views/devise/registrations/_external_openid_connect.html.erb:
```diff
@@ -0,0 +1,21 @@
+<%# If We do have and id we need to present the option to unlink %>
+<% unlinktext = _("Unlink your account from #{scheme.description}. You can link again at any time.") %>
+<% unlinkconf = _("Are you sure you want to unlink #{scheme.description} ID?") %>
+<%= id.value %>
```
Maybe not before the release, but this should be updated at some point. Here's one way to do so:

![Screenshot from 2024-09-06 11-29-34](https://github.com/user-attachments/assets/9eb1a7f7-748a-4524-b6c3-9a875a7460ee)

```
[11, 20] in /home/aaron/Documents/GitHub/roadmap/app/views/devise/registrations/_external_openid_connect.html.erb
   11: <% unlinktext = _("Unlink your account from #{scheme.description}. You can link again at any time.") %>
   12: <% unlinkconf = _("Are you sure you want to unlink #{scheme.description} ID?") %>
   13: <% parsed_attrs = JSON.parse(id.attrs) %>
   14: <% email = parsed_attrs["info"]["email"] %>
   15: <% byebug %>
=> 16: <%= id.value %>
   17: <%= link_to '<i class="fas fa-fw fa-circle-xmark" aria-hidden="true"></i>'.html_safe,
   18:                 destroy_user_identifier_path(id),
   19:                 method: :delete,
   20:                 title: unlinktext,
(byebug) id.value
"http://cilogon.org/serverE/users/123445"
(byebug) email
"askiba@ualberta.ca"
(byebug) 
```
Rather than the current URL, rendering the email value would probably be a lot more useful to users.

lagoan replied:
This will be more usable for users. However, at this time I would like to take more time to consider parsing the attributes, since we cannot  guaranteed that the email will be there. We will follow up this work on #887

aaronskiba replied:
Sounds good. I also added a comment to that issue regarding the wording of "Institutional credentials": https://github.com/portagenetwork/roadmap/issues/887#issuecomment-2341195768

aaronskiba commented on app/views/devise/registrations/_external_openid_connect.html.erb:
```diff
@@ -0,0 +1,21 @@
+<% else %>
+<%# If We do have and id we need to present the option to unlink %>
+<% unlinktext = _("Unlink your account from #{scheme.description}. You can link again at any time.") %>
+<% unlinkconf = _("Are you sure you want to unlink #{scheme.description} ID?") %>
```
Same issue as https://github.com/portagenetwork/roadmap/pull/872/files#r1739377043

200455939-yashu committed:
Merge pull request #882 from portagenetwork/yashu-sso-about-help-updates

About and Help page changes related to the SSO updates #877

aaronskiba commented on CHANGELOG.md:
```diff
@@ -20,6 +22,8 @@
 
 ## [4.1.1+portage-4.1.3] - 2024-08-08
 
+ - Added SSO changes updates to the About and Help pages. [#822](https://github.com/portagenetwork/roadmap/pull/882)
```
Will need to move this to `## [Unreleased]`. Also, `[#822]` should be `[#882]`.

lagoan committed:
Address comments on PR

lagoan committed:
Merge branch 'yashu-sso-link-accounts' of repo

lagoan committed:
Update CHANGELOG

lagoan committed:
Update CHANGELOG

200455939-yashu committed:
Review changes 2 O

200455939-yashu committed:
Resolving the conflicts after merging the link accounts branch

aaronskiba commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -12,6 +12,57 @@ class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+        # If email is missing we need to request the user to register with DMP.
+        # User email can be missing if the usFFvate or trusted clients only we won't get the value.
+        # User email id is one of the mandatory field which is must required.
+        flash[:notice] = _('Something went wrong, Please try signing-up here.')
```
We just need to modify `signing-up` to `signing up`
https://github.com/portagenetwork/roadmap/issues/888#issuecomment-2338911986

aaronskiba commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -12,6 +12,57 @@ class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+                                                      attrs: auth,
+                                                      identifiable: current_user)
+
+        flash[:notice] = _('Linked succesfully')
```
`succesfully` should be `successfully`
https://github.com/portagenetwork/roadmap/issues/888#issuecomment-2338911986

aaronskiba replied:
Actually, I see this has been corrected in app/controllers/users/omniauth_callbacks_controller.rb:52 of https://github.com/portagenetwork/roadmap/pull/869/files

aaronskiba replied:
This will be resolved once #869 is merged

aaronskiba commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -12,6 +12,57 @@ class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+
+    def shibboleth
+      handle_omniauth(IdentifierScheme.for_authentication.find_by(name: 'shibboleth'))
+    end
```
Is this redundant? It looks like the following code might already doing this: https://github.com/portagenetwork/roadmap/pull/872/files#diff-2c20e0c8e95d6a702949c43e99a8de55792f589c286df25403588a5e2e4120b0R6-R13

aaronskiba replied:
Before and after adding `IdentifierScheme.create(name: 'openid_connect', description: "CILogon", active: true, context: 1, identifier_prefix: "http://cilogon.org/serverE/users/")`
```
BEFORE
3.1.4 :001 > IdentifierScheme.for_authentication.pluck(:name)
   (1.8ms)  SELECT "identifier_schemes"."name" FROM "identifier_schemes" WHERE (("identifier_schemes"."context" in (1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31)))
 => ["orcid", "shibboleth"] 
```

```
AFTER
3.1.4 :001 > IdentifierScheme.for_authentication.pluck(:name)
   (1.8ms)  SELECT "identifier_schemes"."name" FROM "identifier_schemes" WHERE (("identifier_schemes"."context" in (1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31)))
 => ["orcid", "shibboleth", "openid_connect"] 
```

aaronskiba replied:
This file contains two methods for each of "orcid", "shibboleth",  and "openid_connect"... One is dynamically created via https://github.com/portagenetwork/roadmap/pull/872/files#diff-2c20e0c8e95d6a702949c43e99a8de55792f589c286df25403588a5e2e4120b0R6-R13 and the other is defined explicitly for each.

aaronskiba commented on app/views/devise/registrations/_personal_details.html.erb:
```diff
@@ -66,30 +66,22 @@
           <span class="aria-only" aria-hidden="false"><%= _('Institutional credentials') %></span>
         </label>
-      <% elsif scheme.name.downcase == 'orcid' %>
-        <%= label_tag(:scheme_name, 'ORCID', class: 'control-label') %>
```
Are we wanting to remove the linking via ORCID? This linking functionality currently works in production.
`integration`
![Screenshot from 2024-09-10 09-26-33](https://github.com/user-attachments/assets/68ccb013-8d22-4930-91ae-980e6032273c)
`yashu-sso-link-accounts`
![Screenshot from 2024-09-10 09-27-41](https://github.com/user-attachments/assets/c481379b-5ed4-4257-9bfd-a45886f3be92)

aaronskiba replied:
Linked account on production site:
<img width="213" alt="Screenshot 2024-09-10 at 9 57 50 AM" src="https://github.com/user-attachments/assets/cbd187ce-305b-4d0c-999c-766af4db75a7">

aaronskiba replied:
Maybe we should've accessed the "openid_connect" IdentifierScheme entry via (the now deleted) `@identifier_schemes`?
https://github.com/portagenetwork/roadmap/pull/872/files#diff-9be479acf1db667ae5055f7c33373128567ea22ee7a11aeff790df6321d54f3bL69

However, `@identifier_schemes` is defined in `app/controllers/registrations_controller.rb` as follows:
```ruby
@identifier_schemes = IdentifierScheme.for_users.order(:name)
```
So we'd also have to set the `context` value of `openid_connect` to also be true for the `for_users` check (change value from 1 to 9).

```ruby
# app/models/identifier_scheme.rb
  has_flags 1 => :for_authentication,
            2 => :for_orgs,
            3 => :for_plans,
            4 => :for_users,
            5 => :for_contributors,
            column: 'context',
```

aaronskiba replied:
We deleted the dynamic methods, this is now resolved.

200455939-yashu committed:
Review changes 2.1

200455939-yashu committed:
Removing conflict HEAD

aaronskiba committed:
Patch to put back ORCID linking functionality

Commit f76758ba7a7e88b22eb74c7432e00bca5a5c3d84 - "Add link account with CILogon" added the functionality to link an external account via CILogon. However, the commit also removed the functionality to link one's ORCID credentials within the app. This commit adds the ORCID functionality back.

lagoan committed:
Fix broken test

lagoan committed:
Remove unused code in omniauth callbacks

We previously had methods that were dynamically created but are no
longer needed since we are creating them specifically.

This is done to handle the different workflow for CILogon account
creation.

aaronskiba committed:
Make rubocop happy

aaronskiba committed:
Update CHANGELOG.md

aaronskiba committed:
Merge pull request #891 from portagenetwork/aaron/issues/orcid-patch

Patch to Put Back ORCID Linking Functionality

aaronskiba committed:
Merge pull request #869 from portagenetwork/yashu-controller-spec

Testcases for the Ominiauth controller openid connect

aaronskiba committed:
Remove hyphen from "signing-up" string

Change was requested here: https://github.com/portagenetwork/roadmap/issues/888#issuecomment-2338911986

aaronskiba committed:
Add rake task for `4.1.1+portage-4.2.0` upgrade

Release `4.1.1+portage-4.2.0` includes a migration and it requires a specific new entry to the `identifier_schemes` table for openid_connect / CILogon to work properly. This rake task is intended to take care of all of that.

Note: The changes applied via this rake task have already been executed manually on the staging environment. However, performing the rake task on staging should still have no adverse effects.

aaronskiba committed:
Update CHANGELOG.md

aaronskiba committed:
Merge pull request #892 from portagenetwork/aaron/add-4.2.0-rake-upgrade-task

Add Rake Task For `4.1.1+portage-4.2.0` Upgrade

aaronskiba reviewed (approved):
Approving this so it can be merged into integration for full testing of upcoming release.
